File: JavaScript/array/queue-circular.js
/**
 * Author: Axel Ariel Saravia
 */

/**
FixedArray<T> :: {
    length: number [uint]
    [0..]: T
}

*/
/**
CircularQueue<T> :: {
    capacity: number [uint > 0],
    content: FixedArray<T>,
    head: number [uint],
    length: number [uint]
}
*/

/**
createFixedArray :: (number [uint > 0]) -> FixedArray<T> */
function createFixedArray(capacity) {
    if (!Number.isSafeInteger(capacity) || capacity < 1) {
        throw new Error("capacity is not a unsigned integer");
    }
    const arr = {
        length: capacity,
    };
    for (let i = 0; i < capacity; i += 1) {
        arr[i] = undefined;
    }
    return Object.seal(arr);
}

const CircularArrayQueue = Object.freeze({
    /**
    create :: (number [uint > 0]) -> CircularQueue<T> */
    create(capacity) {
        return Object.seal({
            capacity: capacity,
            content: createFixedArray(capacity),
            head: 0,
            length: 0
        });
    },
    /**
    add :: (CircularQueue<T>, T) -> boolean */
    add(queue, x) {
        if (queue.length !== queue.capacity) {
            queue.content[queue.head] = x;
            queue.head = (queue.head + 1) % queue.capacity;
            queue.length += 1;
            return true;
        } else {
            return false;
        }
    },
    /**
    remove :: (CircularQueue<T>) -> boolean */
    remove(queue) {
        if (queue.length > 0) {
            queue.length -= 1;
            return true;
        } else {
            return false;
        }
    }
});

export default CircularArrayQueue;

File: C/array/queue-circular.c
/**
 * Author: Axel Ariel Saravia
 */

# include "queue-circular.h"

circularQueue circular_queue_init(size_t const capacity, T arr[capacity]) {
    return (circularQueue){
        .content = arr,
        .capacity = capacity,
        .length = 0,
        .head = 0,
    };
}

unsigned circular_queue_remove(circularQueue cq[static 1]) {
    unsigned ret = 0;
    if (cq->length) {
        ret = cq->content[cq->head];
        cq->length -= 1;
    }
    return ret;
}

circularQueue* circular_queue_add(circularQueue cq[static 1], unsigned value) {
    if (cq->length != cq->capacity) {
        cq->content[cq->head] = value;
        *cq = (circularQueue){
            .content = cq->content,
            .capacity = cq->capacity,
            .length = cq->length + 1,
            .head = (cq->head + 1) % cq->capacity,
        };
    }
    return cq; 
}
